common/autoware_universe_utils/src/geometry/ear_clipping.cpp

```diff
@@ -20,311 +20,392 @@
 namespace autoware::universe_utils
 {
 
-void EarClipping::operator()(const Polygon2d & polygon)
+std::size_t construct_point(std::size_t index, const Point2d & point, std::vector<Point> & points)
 {
-  indices.clear();
-  vertices = 0;
-
-  if (polygon.outer().size() == 0) return;
-
-  std::size_t len = 0;
+  points.emplace_back(index, point);
+  return index;
+}
 
+std::vector<Point> perform_triangulation(
+  const Polygon2d & polygon, std::vector<std::size_t> & indices)
+{
+  indices.clear();
+  std::vector<Point> points_;
+  std::size_t vertices = 0;
   const auto & outer_ring = polygon.outer();
-  len = outer_ring.size();
+  std::size_t len = outer_ring.size();
   points_.reserve(len * 3 / 2);
+
+  if (polygon.outer().empty()) return points_;
+
   indices.reserve(len + outer_ring.size());
+  auto outer_point = linked_list(outer_ring, true, vertices, points_);
+  if (outer_point == points_[outer_point.value()].prev_index.value()) {
```

Comment:
Is it possible that `prev_index` is `nullopt` ? in which case this should be checked.
What about `outer_point` ? If not it should not be a `std::optional`.